### PR TITLE
DL3-34 Course Picker: Add Pagination to the Select Course UI

### PR DESCRIPTION
**Pull request by mpellicer**
The Harmony courses endpoint supports pagination returning a metadata object in the body, both the frontend and the backend must support the new schema and consider paginated requests.

## Description
Includes support for pagination in the frontend and the backend, by default the front requests the first page and you can navigate through pages using the Harmony endpoint.

### Motivation and Context
The Harmony courses endpoint supports pagination.

### Fixes
[Jira ticket](https://lumenlearning.atlassian.net/browse/DL3-34)

## How Has This Been Tested?
This has been tested using static data in the local development environment, mocking data using D2L and the middleware, and with the real endpoint response using D2L and the middleware.

Also tested the real endpoint using Canvas.

## Screenshots

pagination_real_endpoint
pagination_fake_endpoint
pagination_a11y

Using the real endpoint and Canvas:

pagination_canvas

---

## New ENV variables

---

## Checklist

- [X] I have reviewed the AC for this feature and my changes meet all requirements
- [X] (For UI changes) I have considered the accessibility of this feature (see https://www.a11yproject.com/checklist/ for a high-level checklist)
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes
- [X] Any dependent changes have been merged and published in downstream modules
- [X] I have requested a review from at least one other dev

### CODE REVIEW

File: src/main/frontend/src/app/appSlice.js [hunk 1]
```diff
@@ -34,29 +27,39 @@ export const appSlice = createSlice({
...
 export const selectClientId = (state) => state.clientId;
 export const selectDeploymentId = (state) => state.deploymentId;
 export const selectIdToken = (state) => state.id_token;
 export const selectLoading = (state) => state.loading;
+export const selectErrorFetchingCourses = (state) => state.errorFetchingCourses;
 
 // This function fetches the courses from the backend, it should be invoked when loading the application.
-export const fetchCourses = (state) => (dispatch) => {
+export const fetchCourses = (page) => (dispatch, getState) => {
+  // Get the id_token from the state, thunks already have access to getState as the second argument.
```
Comment (mgwozdz-unicon) [inline src/main/frontend/src/app/appSlice.js:53]: Nitpick: I think "thunks" is a typo

File: src/main/java/net/unicon/lti/service/harmony/HarmonyService.java [hunk 2]
```diff
@@ -39,19 +41,31 @@ public class HarmonyService {
...
+                // Add the page parameter since this API is paginated.
+                builder.setParameter("page", String.valueOf(page));
+                requestUrl = builder.build().toString();
+            }
+
             HttpHeaders headers = new HttpHeaders();
             headers.setBearerAuth(harmonyJWT);
             HttpEntity<String> entity = new HttpEntity<>("parameters", headers);
-            ResponseEntity<HarmonyPageResponse> response = restTemplate.exchange(harmonyCoursesApiUrl, HttpMethod.GET, entity, HarmonyPageResponse.class);
+            ResponseEntity<HarmonyPageResponse> response = restTemplate.exchange(requestUrl, HttpMethod.GET, entity, HarmonyPageResponse.class);
```
Comment (mgwozdz-unicon) [inline src/main/java/net/unicon/lti/service/harmony/HarmonyService.java:68]: Could you please add a unit test to verify that if page equals 2, then the requestUrl value has the page parameter added to it as expected?

**Review by mgwozdz-unicon — CHANGES_REQUESTED** [review]

File: src/main/frontend/src/app/appSlice.js (hunk 1 above)
Comment (mpellicer) [reply src/main/frontend/src/app/appSlice.js:53]: I think it's correct Mary, it's a thunk function.

https://redux.js.org/usage/writing-logic-thunks

File: src/main/java/net/unicon/lti/service/harmony/HarmonyService.java (hunk 2 above)
Comment (mpellicer) [reply src/main/java/net/unicon/lti/service/harmony/HarmonyService.java:68]: Makes sense to me, I created a unit test to verify both the page number and the 'page' parameter name.

**Review by smith750 — APPROVED** [review]
This looks great - thank you!

**Pull request merged**